docs/design/2020-05-08-cluster-index.md

```diff
@@ -0,0 +1,254 @@
+
+So if a query doesn't reference column `e`, the index `c_d` is a covering index, we can build an IndexScan plan
+instead of an IndexLookUp plan.
+
+### Coprocessor
+Besides codec code, several coprocessor executors need to be changed to support common handle.
+
+#### IndexScan
+If the IndexScan schema doesn't contain primary key columns, the coprocessor logic is unchanged.
+
+If the IndexScan schema contains primary key columns, the coprocessor should decode the common handle
+to primary key column values.
+
+The encoded column value can be obtained by the `EncodedCol` method of the handle.
+
+#### Analyze Column
+If the primary key is multi-column, CMSketch needs to insert prefix column values of the primary key. 
+
+#### Fast Analyze
+Since the handle is not integer anymore, we cannot randomly generate the sample key at a given position.
+So we need to add a coprocessor request type that samples the key/value pair during the scan and return the statistics result.
+
+### Executors
+Several executors need to be changed to support common handle
+
+#### Admin Executors
+The `admin check/recover table/index` executors heavily depend on the assumption that handle is int64 handle, 
+it needs to do major changes to support common handle.
+
+#### PointGet/BatchPointGet
+We need to decode the common handle to chunk. 
+
+#### Insert
+We don't need to encode primary columns into the row value.
+
+#### Update
+If any of the primary key columns changed, we need to delete the row and insert the row.
+
+#### SplitTableRegion
+We need to redesign this executor since we cannot find a middle key based on the region start/end key. 
+
+## Rationale
+
+The clustered index feature would remove the extra read/write cost for tables with a non-single-integer-column primary key.
+An experiment on the tpcC workload by compressing multi-column PK into one integer PK shows a 33% improvement. 
+
+## Compatibility
+
+This feature doesn't break the backward compatibility.
+
+We add a field `IsCommonHandle` in the `model.TableInfo` and a global variable `tidb_enable_clustered_index`.
+
+When creating a new table, set this field to true if the global variable `tidb_enable_clustered_index` is set to `1`.
+
+All the common handle related logic should check `IsCommonHandle` field before entering the common handle code block.
+
+For a new cluster, we insert the global variable `tidb_enable_clustered_index` value `1` in the bootstrap phase, the clustered index feature is supported by default.
+
+For old clusters upgrade to the latest release version, the user needs to explicitly set the global variable to `1` to use the clustered index feature.
+
+## Implementation
+
+```
+implement the Handle interface for IntHandle and CommonHandle.
+  |
+  + refactor int64 handle to the Handle interface in all tidb packages.
+  |
+  +-+ implement the codec.
+    |
+    |-+ support codec in tikv.
```

Comment:
maybe cdc/tiflash/tispark/binlog also need support this, because row key changed and row values didn't contain primary key columns

Reply:
Yes, this should be added to the open issue section in the proposal.